### ISSUE TITLE
Much higher IOPS results in random read test than other tools

### ISSUE THREAD

**Issue by 111alan** [microsoft/diskspd#131]
I ran diskspd with these commands, to test out the storage system performance(4KB random read, QD16x16T) of different hardware platforms.
`diskspd.exe -b4K -t16 -r -o16 -d39 -Sh E:\iobw.tst`

Ran on EPYC2-7702 and PM1725a, with a 20GB prebuilt file with pseudo random content created with iometer. All 3 tests are done with the same file. But the results is way higher than the number given by fio and iometer, which both offer the similiar results, about 400K iops. But on intel platform with xeon platinum 8275CL cpu the results are quite consistent(all tests are about 800K iops).

I tried to create a new file for testing, but all 3 tests produced the same results as before. What could go wrong? Thank you.
![7702_windows_5x](https://user-images.githubusercontent.com/30613544/75492178-2ff89780-59f2-11ea-8d27-b439a4b931af.jpg)

**Comment by 111alan** [comment]
BTW there's a similiar result by one of my friends using a PM981 512GB on a EPYC2 7502 platform, which was rated about 440K iops, but the result is also much higher(583K).
`C:\Windows\system32>diskspd.exe -b4K -t8 -r -o32 -d16 -Sh #1

Command Line: diskspd.exe -b4K -t8 -r -o32 -d16 -Sh #1

Input parameters:

        timespan:   1
        -------------
        duration: 16s
        warm up time: 5s
        cool down time: 0s
        random seed: 0
        path: '#1'
                think time: 0ms
                burst size: 0
                software cache disabled
                hardware write cache disabled, writethrough on
                performing read test
                block size: 4096
                using random I/O (alignment: 4096)
                number of outstanding I/O operations: 32
                thread stride size: 0
                threads per file: 8
                using I/O Completion Ports
                IO priority: normal

System information:

        computer name: EPYC
        start time: 2020/02/04 08:37:00 UTC

Results for timespan 1:
*******************************************************************************

actual test time:       16.00s
thread count:           8
proc count:             64

Read IO
thread |       bytes     |     I/Os     |    MiB/s   |  I/O per s |  file
------------------------------------------------------------------------------
     0 |      3172945920 |       774645 |     189.12 |   48414.09 | #1 (476GiB)
     1 |      3180544000 |       776500 |     189.57 |   48530.02 | #1 (476GiB)
     2 |      3195088896 |       780051 |     190.44 |   48751.95 | #1 (476GiB)
     3 |      3187572736 |       778216 |     189.99 |   48637.27 | #1 (476GiB)
     4 |      6368505856 |      1554811 |     379.58 |   97173.23 | #1 (476GiB)
     5 |      6369304576 |      1555006 |     379.63 |   97185.41 | #1 (476GiB)
     6 |      6366236672 |      1554257 |     379.45 |   97138.60 | #1 (476GiB)
     7 |      6368272384 |      1554754 |     379.57 |   97169.66 | #1 (476GiB)
------------------------------------------------------------------------------
total:       38208471040 |      9328240 |    2277.34 |  583000.23`

**Comment by dl2n** [comment]
I'll think about the fist case in a bit.

Re: the second case, a suspect pops out quickly.

The [PM981](https://www.samsung.com/semiconductor/ssd/client-ssd/MZVLB512HAJQ/) specification from Samsung shows 270K random read and 420K random write. The type of IO is very important. You're showing a random read result at 583K, which is then way higher (>2x) than Samsung's spec.

As to that spec: Samsung is providing no context on test conditions, but given the difference in behavior there is a clear suspect - device preconditioning and the amount/structure of the data that has been written to the device prior to the measured run.

What's probably happening is that the SSD has had very little data written to it (relative to capacity) since it was last cleared and the device can optimize away reading the SSD's NAND since it knows it hasn't been written - it just returns zeroes. This can obviously be done a lot faster since its just stressing NVMe, the speed of the device's FTL processor and transfers over the PCIe bus attachment. If you were to precondition the device with writes (random is usually the choice since that will randomize the SSD's physical<->logical block mapping tables) the read performance would probably settle close to where Samsung documents it: the data would actually be coming all the way from NAND.

Even if the SSD happens to be formatted with a filesystem and some data has been written to it (say, even an OS installation), its still quite possible for a fresh system device to have only had a small fraction of its capacity touched by the time of this test (you tell me?). If say only 50GB of data has been written out of the device's 512GB capacity (476GiB), ~90% of the reads could be optimized away.

The difference between IOMETER, FIO and DISKSPD in this case may simply be down to a somewhat more Windows-native IO path in DISKSPD's case.

Some references you can dip into, and/or simply search for "ssd device preconditioning" and you'll find a lot of discussion.

[SNIA SDC 2011 talk slides](https://www.snia.org/sites/default/education/tutorials/2011/fall/SolidState/EstherSpanjer_The_Why_How_SSD_Performance_Benchmarking.pdf)
[SNIA SSD Testing Specification](https://www.snia.org/tech_activities/standards/curr_standards/pts)

Bottom line is that this isn't simple on modern devices. It is easy to fall into behavioral traps or, in another sense, learn a lot about how your system and storage works.

**Comment by 111alan** [comment]
> I'll think about the fist case in a bit.
> Re: the second case, a suspect pops out quickly.
> The PM981 specification from Samsung shows 270K random read and 420K random write. The type of IO is very important. You're showing a random read result at 583K, which is then way higher (>2x) than Samsung's spec.
> As to that spec: Samsung is providing no context on test conditions, but given the difference in behavior there is a clear suspect - device preconditioning and the amount/structure of the data that has been written to the device prior to the measured run.
> What's probably happening is that the SSD has had very little data written to it (relative to capacity) since it was last cleared and the device can optimize away reading the SSD's NAND since it knows it hasn't been written - it just returns zeroes. This can obviously be done a lot faster since its just stressing NVMe, the speed of the device's FTL processor and transfers over the PCIe bus attachment. If you were to precondition the device with writes (random is usually the choice since that will randomize the SSD's physical<->logical block mapping tables) the read performance would probably settle close to where Samsung documents it: the data would actually be coming all the way from NAND.
> Even if the SSD happens to be formatted with a filesystem and some data has been written to it (say, even an OS installation), its still quite possible for a fresh system device to have only had a small fraction of its capacity touched by the time of this test (you tell me?). If say only 50GB of data has been written out of the device's 512GB capacity (476GiB), ~90% of the reads could be optimized away.
> The difference between IOMETER, FIO and DISKSPD in this case may simply be down to a somewhat more Windows-native IO path in DISKSPD's case.
> Some references you can dip into, and/or simply search for "ssd device preconditioning" and you'll find a lot of discussion.
> SNIA SDC 2011 talk slides
> SNIA SSD Testing Specification
> Bottom line is that this isn't simple on modern devices. It is easy to fall into behavioral traps or, in another sense, learn a lot about how your system and storage works.

I actually have the same thought, that's one of the reason why I insist to test the drive AFTER formating and preconditioning, and test the file directly. Because even you've preconditioned the RAW drive in advance the SSD's GC mechanics could still get in the way. My friend seems to be too busy to do that so I did the test myself, that spawns the first result.

**Comment by dl2n** [comment]
OK, lets consider the second case resolved then.

The specifics of preconditioning can be all important (they were in the other issue you initially commented on, too). Can you describe what you do?

Your comment of "preconditioned the RAW drive in advance" ... and then formatting with a filesystem raises the potential of a big trap. It won't affect your read-after-write test ordering, but you should be aware that formatting re-TRIM's the device, tossing out all of the lba-pba preconditioning.

As far as the (PM1725a)[https://www.samsung.com/semiconductor/global.semi.static/Samsung_PM1725a_NVMe_SSD-0.pdf], the 2.5" U.2 form factor is rated for 800K random read. I'm not sure DISKSPD getting closer to the rated device capability than IOMETER or FIO is strictly speaking strange - I'm confident DISKSPD is actually running that IO in and out of the device. It may be some aspect of the more Windows-native IO path in DISKSPD relative to them.

If this is all side-by-side one after the other runs on the same file, the couple ideas I have are related to worker thread affinity and alternation on NUMA domains. How many logical numa nodes does the EPYC2-7702 expose? See if FIO/IOMETER exercise the same one run to run, they may alternate if they aren't being careful (at least by default) about how the platform will round-robin processes between NUMA domains to level load around the system complex. DISKSPD defaults to hard thread affinitization starting at socket 0 core 0, and ascending through the topology from there.

If the Intel processor only enumerates as a single NUMA it might suggest something about the relative attachment of the PM1725a to the topology, or some other aspect of the AMD design. I'm not immediately familiar with what that could be.

**Comment by dl2n** [comment]
p.s. I don't see the 8275 (CL = Cascade Lake?) at [Intel Ark](https://ark.intel.com/content/www/us/en/ark/products/codename/124664/cascade-lake.html) spec site. 8270 and three variants of 8276 (8276, 8276M and 8276L).

**Comment by 111alan** [comment]
> OK, lets consider the second case resolved then.
> The specifics of preconditioning can be all important (they were in the other issue you initially commented on, too). Can you describe what you do?
> Your comment of "preconditioned the RAW drive in advance" ... and then formatting with a filesystem raises the potential of a big trap. It won't affect your read-after-write test ordering, but you should be aware that formatting re-TRIM's the device, tossing out all of the lba-pba preconditioning.
> As far as the (PM1725a)[https://www.samsung.com/semiconductor/global.semi.static/Samsung_PM1725a_NVMe_SSD-0.pdf], the 2.5" U.2 form factor is rated for 800K random read. I'm not sure DISKSPD getting closer to the rated device capability than IOMETER or FIO is strictly speaking strange - I'm confident DISKSPD is actually running that IO in and out of the device. It may be some aspect of the more Windows-native IO path in DISKSPD relative to them.
> If this is all side-by-side one after the other runs on the same file, the couple ideas I have are related to worker thread affinity and alternation on NUMA domains. How many logical numa nodes does the EPYC2-7702 expose? See if FIO/IOMETER exercise the same one run to run, they may alternate if they aren't being careful (at least by default) about how the platform will round-robin processes between NUMA domains to level load around the system complex. DISKSPD defaults to hard thread affinitization starting at socket 0 core 0, and ascending through the topology from there.
> If the Intel processor only enumerates as a single NUMA it might suggest something about the relative attachment of the PM1725a to the topology, or some other aspect of the AMD design. I'm not immediately familiar with what that could be.

Both processors are running with 1 numa(AMD NPS1) mode, the EPYC2 7702 have 2 processor groups in windows, but the test threads are strictly running in the first group during the all the read tests, as shown in the task manager.

By preconditioning the raw drive I also mean run the write i/os directly into the raw drive for some time, in this case the drive may still think these are useless data and run garbage collection over them.

**Comment by 111alan** [comment]
> p.s. I don't see the 8275 (CL = Cascade Lake?) at Intel Ark spec site. 8270 and three variants of 8276 (8276, 8276M and 8276L).

correction: after update the 7702 at NPS1 is running with a single processor group in windows.

8275CL is a CLX 24C48T processor with 3.0 base clock and 3.6-3.9 turbo and 240w tdp.
![7702_util](https://user-images.githubusercontent.com/30613544/75395417-53a5da00-592d-11ea-8569-7fa6b4b33d64.png)

**Comment by dl2n** [comment]
Try to get to this view.

You'll see my 2950x surfaces 1 socket but 2 NUMA domains. From the hot cores in your Task Manager view, I strongly suspect you have 2 as well and its running in the second on this run. It may alternate back to the first if you stop/restart IOMETER/dynamo.

![image](https://user-images.githubusercontent.com/8743209/75395665-b48acf00-58a7-11ea-81b3-aafe29c26392.png)

**Comment by 111alan** [comment]
> Try to get to this view.
> You'll see my 2950x surfaces 1 socket but 2 NUMA domains. From the hot cores in your Task Manager view, I strongly suspect you have 2 as well and its running in the second on this run. It may alternate back to the first if you stop/restart IOMETER/dynamo.

Indeed there are still 2 groups. Tried directly binding the core affinity to cpu 0-31, the results are still almost the same.

By the way, by setting the ioengine to windowsaio the FIO should also use windows-native io path right?

**Comment by 111alan** [comment]
I just noticed that on intel platform the result generated by diskspd is a bit lower than the other 2(this is consistent).
![8275CL_windows](https://user-images.githubusercontent.com/30613544/75400746-1e07ed80-593b-11ea-8b64-2aeb6ead8127.JPG)

**Comment by ale-goncalves** [comment]
Testing with enterprise level SSD, besides preconditioning the device to
achieve a steady state, there is the amount of DRAM cache inside the
device. Which will interfere in write testing mainly. So need to be aware
of that as well.
On the CPU side EPYC added different scheme for core to cache relation, and
it is also different on Gen1 and Gen2 EPYC system. On Zeppelin die the LLC
relation is different from XEON.
[image: image.png]

EPYC Naples introduced a 1:4:2 model, one socket, that contains four NUMA
domains, and each NUMA domain contains two LLC domains. Rome provides a
NUMA model similar to the XEON, with a single socket and single NUMA
domain. However, each chiplet contains two separate LLC domains. A Rome CPU
package contains eight chiplets, and thus, 16 different LLC domains exist
within a socket & NUMA domain.

So thread scheduling will have a bigger impact when moving from cores in
the same NUMA but with different CPU cache lines.

**Comment by 111alan** [comment]
> Testing with enterprise level SSD, besides preconditioning the device to
> achieve a steady state, there is the amount of DRAM cache inside the
> device. Which will interfere in write testing mainly. So need to be aware
> of that as well.
> On the CPU side EPYC added different scheme for core to cache relation, and
> it is also different on Gen1 and Gen2 EPYC system. On Zeppelin die the LLC
> relation is different from XEON.
> [image: image.png]
> 
> EPYC Naples introduced a 1:4:2 model, one socket, that contains four NUMA
> domains, and each NUMA domain contains two LLC domains. Rome provides a
> NUMA model similar to the XEON, with a single socket and single NUMA
> domain. However, each chiplet contains two separate LLC domains. A Rome CPU
> package contains eight chiplets, and thus, 16 different LLC domains exist
> within a socket & NUMA domain.
> 
> So thread scheduling will have a bigger impact when moving from cores in
> the same NUMA but with different CPU cache lines.
> …

I was only testing read performance, so there shouldn't be a drastic performance drop when entering transition state or steady state like a write test does.

I actually did a research on EPYC2 Rome's interconnection architecture, you can see the article with the conceptual model here:
[http://www.ssdfans.com/blog/2020/02/22/epyc2-rome拓扑结构测试分析三大方法/](url)

The intention of me binding the test tools to cpu 0-31 is to only use the first node in the north bridge. And the highly used cores didn't alternate between cores at all. That still doesn't explain the difference between diskspd and the other two tools.

**Comment by dl2n** [comment]
Let me see if I restate where we are right now.

In the initial case DISKSPD is driving significantly higher read IOPS than IOMETER/FIO, closer to the stated capability of the PM1725a on AMD EPYC. This seems like something which comes down to triaging why IOMETER/FIO are running slower, not DISKSPD running faster. I can't speak to the windowsaio support in FIO and whether it matches well with DISKSPD's native implementation but I think this is a great question for the FIO community. DISKSPD's source is right here to compare and reuse as needed under the MIT license. Can you take this in that direction?

In the second case DISKSPD is driving modestly lower on an Intel platform, the 8275CL?

The other thing I noted in the task manager view of the EPYC case was that IOMETER seemed to be dividing affinitized threads in a simultaneous multi-thread (~intel hyper-thread) aware fashion. DISKSPD does not have this intelligence by default, but it is possible to do so with the -ag flag (-ag0,0,2,4,6 ... etc. to assign threads to kgroup 0 core 0, 2, 4, 6 ...). If you do that on the Intel platform there should be some improvement, and you should be able to easily spot how IOMETER/FIO lay out as compared to DISKSPD ahead of time.

Getting this intelligence into DISKSPD either as a new default (some risk in confounding old-new result comparability) or an easily accessible option (not requiring extensive specification of core masks with -ag) has been a thought for a while.

**Comment by 111alan** [comment]
Thx for the reply, seems that there's something deep going on, I'll also post this on fio's page for more discussion.

Yes on intel platform diskspd get a bit lower results and on AMD platform diskspd get a bit higher results, I just disabled the SMT and HT on both processors, the results seem to stay the same.

**Comment by ale-goncalves** [comment]
Just to refresh the thread. Several people have reported this issue to me. Digging a little deeper, at least in FIO for Windows, one of the main reasons is regarding IOCP - IO Completion Ports - which FIO is either not using or creating Blocking situation with a synchronous pattern. To have Diskspd behave like FIO (downgrade performance) we need to use -x switch which does not use IOCP. For the meantime if there is a need to benchmark Storage on Microsoft Windows we should stick with diskspd. Check thread on FIO https://github.com/axboe/fio/issues/1073

**Comment by krisiasty** [comment]
One more thing... usually you should expect higher random read performance than random write, but most (if not all) Samsung Client SSD drives, incl. PM981 mentioned above, use "Intelligent Turbo Write" technology to speed-up short bursts of writes resulting in better write performance in the data sheet (notice they say "up to 420k IOPS" or something like that). In reality the drive is unable to sustain that performance level, so any benchmark longer than a few seconds should show much lower performance.
